File: src/firebase/paymentService.js
import { 
  collection, 
  doc, 
  getDocs, 
  getDoc, 
  addDoc, 
  updateDoc, 
  query, 
  where, 
  orderBy, 
  limit,
  Timestamp
} from 'firebase/firestore';
import { db } from './config';

// Payment status
export const PAYMENT_STATUS = {
  PENDING: 'PENDING',
  COMPLETED: 'COMPLETED',
  FAILED: 'FAILED',
  REFUNDED: 'REFUNDED',
  CANCELLED: 'CANCELLED'
};

// Payment method
export const PAYMENT_METHOD = {
  CASH: 'CASH',
  CREDIT_CARD: 'CREDIT_CARD',
  DEBIT_CARD: 'DEBIT_CARD',
  BANK_TRANSFER: 'BANK_TRANSFER',
  DIGITAL_WALLET: 'DIGITAL_WALLET'
};

// Create payment
export const createPayment = async (paymentData) => {
  try {
    const payment = {
      ...paymentData,
      status: PAYMENT_STATUS.PENDING,
      createdAt: Timestamp.now(),
      updatedAt: Timestamp.now()
    };
    
    const docRef = await addDoc(collection(db, 'payments'), payment);
    
    return {
      success: true,
      data: {
        id: docRef.id,
        ...payment,
        createdAt: payment.createdAt.toDate(),
        updatedAt: payment.updatedAt.toDate()
      }
    };
  } catch (error) {
    console.error('Create payment error:', error);
    return {
      success: false,
      error: 'Failed to create payment.'
    };
  }
};

// Get payment by ID
export const getPaymentById = async (paymentId) => {
  try {
    const docRef = doc(db, 'payments', paymentId);
    const docSnap = await getDoc(docRef);
    
    if (docSnap.exists()) {
      const payment = {
        id: docSnap.id,
        ...docSnap.data(),
        createdAt: docSnap.data().createdAt?.toDate(),
        updatedAt: docSnap.data().updatedAt?.toDate()
      };
      
      return {
        success: true,
        data: payment
      };
    } else {
      return {
        success: false,
        error: 'Payment not found.'
      };
    }
  } catch (error) {
    console.error('Get payment error:', error);
    return {
      success: false,
      error: 'Failed to fetch payment.'
    };
  }
};

// Get payments by order ID
export const getPaymentsByOrder = async (orderId) => {
  try {
    let q = query(
      collection(db, 'payments'),
      where('orderId', '==', orderId),
      orderBy('createdAt', 'desc')
    );
    
    const querySnapshot = await getDocs(q);
    const payments = [];
    
    querySnapshot.forEach((doc) => {
      payments.push({
        id: doc.id,
        ...doc.data(),
        createdAt: doc.data().createdAt?.toDate(),
        updatedAt: doc.data().updatedAt?.toDate()
      });
    });
    
    return {
      success: true,
      data: payments
    };
  } catch (error) {
    console.error('Get payments by order error:', error);
    return {
      success: false,
      error: 'Failed to fetch payments.'
    };
  }
};

// Get payments by customer
export const getPaymentsByCustomer = async (customerId) => {
  try {
    let q = query(
      collection(db, 'payments'),
      where('customerId', '==', customerId),
      orderBy('createdAt', 'desc')
    );
    
    const querySnapshot = await getDocs(q);
    const payments = [];
    
    querySnapshot.forEach((doc) => {
      payments.push({
        id: doc.id,
        ...doc.data(),
        createdAt: doc.data().createdAt?.toDate(),
        updatedAt: doc.data().updatedAt?.toDate()
      });
    });
    
    return {
      success: true,
      data: payments
    };
  } catch (error) {
    console.error('Get payments by customer error:', error);
    return {
      success: false,
      error: 'Failed to fetch payments.'
    };
  }
};

// Update payment status
export const updatePaymentStatus = async (paymentId, status) => {
  try {
    const paymentRef = doc(db, 'payments', paymentId);
    const paymentSnap = await getDoc(paymentRef);
    
    if (!paymentSnap.exists()) {
      return {
        success: false,
        error: 'Payment not found.'
      };
    }
    
    await updateDoc(paymentRef, {
      status,
      updatedAt: Timestamp.now()
    });
    
    return {
      success: true,
      message: 'Payment status updated successfully.'
    };
  } catch (error) {
    console.error('Update payment status error:', error);
    return {
      success: false,
      error: 'Failed to update payment status.'
    };
  }
};

// Process payment (simplified - in real app, integrate with payment gateway)
export const processPayment = async (paymentId) => {
  try {
    // Simulate payment processing
    // In a real application, you would integrate with a payment gateway
    await new Promise(resolve => setTimeout(resolve, 1000));
    
    // For demo purposes, we'll mark it as completed
    return await updatePaymentStatus(paymentId, PAYMENT_STATUS.COMPLETED);
  } catch (error) {
    console.error('Process payment error:', error);
    return {
      success: false,
      error: 'Failed to process payment.'
    };
  }
};

// Get payment statistics
export const getPaymentStatistics = async (customerId = null) => {
  try {
    let q = collection(db, 'payments');
    
    if (customerId) {
      q = query(q, where('customerId', '==', customerId));
    }
    
    const querySnapshot = await getDocs(q);
    const payments = [];
    
    querySnapshot.forEach((doc) => {
      payments.push(doc.data());
    });
    
    const stats = {
      total: payments.length,
      totalAmount: payments.reduce((sum, payment) => sum + payment.amount, 0),
      pending: payments.filter(payment => payment.status === PAYMENT_STATUS.PENDING).length,
      completed: payments.filter(payment => payment.status === PAYMENT_STATUS.COMPLETED).length,
      failed: payments.filter(payment => payment.status === PAYMENT_STATUS.FAILED).length,
      refunded: payments.filter(payment => payment.status === PAYMENT_STATUS.REFUNDED).length
    };
    
    return {
      success: true,
      data: stats
    };
  } catch (error) {
    console.error('Get payment statistics error:', error);
    return {
      success: false,
      error: 'Failed to fetch payment statistics.'
    };
  }
}; 

File: emart-backend/src/main/java/com/emart/service/PaymentService.java
package com.emart.service;

import com.emart.model.Payment;
import com.emart.model.Order;
import com.emart.model.Invoice;
import com.emart.repository.PaymentRepository;
import com.emart.repository.OrderRepository;
import com.emart.repository.InvoiceRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
public class PaymentService {
    
    private final PaymentRepository paymentRepository;
    private final OrderRepository orderRepository;
    private final InvoiceRepository invoiceRepository;
    private final OrderService orderService;
    private final InvoiceService invoiceService;
    
    @Transactional
    public Payment processPayment(String orderId, Payment.PaymentMethod paymentMethod, 
                                BigDecimal amount, String gatewayResponse) {
        // Validate order exists
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("Order not found"));
        
        // Create payment record
        Payment payment = new Payment();
        payment.setTransactionId(generateTransactionId());
        payment.setOrderId(orderId);
        payment.setOrderNumber(order.getOrderNumber());
        payment.setCustomerId(order.getCustomerId());
        payment.setCustomerName(order.getCustomerName());
        payment.setCustomerEmail(order.getCustomerEmail());
        payment.setAmount(amount);
        payment.setPaymentMethod(paymentMethod);
        payment.setGatewayResponse(gatewayResponse);
        payment.setGatewayTransactionId(generateGatewayTransactionId());
        payment.setStatus(Payment.PaymentStatus.PENDING);
        payment.setPaymentDate(LocalDateTime.now());
        payment.setCreatedAt(LocalDateTime.now());
        payment.setUpdatedAt(LocalDateTime.now());
        
        // Save payment
        Payment savedPayment = paymentRepository.save(payment);
        
        // Process payment (simulate payment gateway)
        boolean paymentSuccess = processPaymentWithGateway(payment);
        
        if (paymentSuccess) {
            // Update payment status
            savedPayment.setStatus(Payment.PaymentStatus.COMPLETED);
            savedPayment.setUpdatedAt(LocalDateTime.now());
            paymentRepository.save(savedPayment);
            
            // Update order payment status
            orderService.updatePaymentStatus(orderId, Order.PaymentStatus.PAID);
            
            // Update invoice status if exists
            try {
                Invoice invoice = invoiceService.getInvoiceByOrderId(orderId);
                invoiceService.updateInvoiceStatus(invoice.getId(), Invoice.InvoiceStatus.PAID);
            } catch (Exception e) {
                log.warn("No invoice found for order: {}", orderId);
            }
            
            log.info("Payment processed successfully for order: {}", orderId);
        } else {
            // Payment failed
            savedPayment.setStatus(Payment.PaymentStatus.FAILED);
            savedPayment.setFailureReason("Payment gateway processing failed");
            savedPayment.setUpdatedAt(LocalDateTime.now());
            paymentRepository.save(savedPayment);
            
            log.error("Payment failed for order: {}", orderId);
        }
        
        return savedPayment;
    }
    
    public Payment getPaymentById(String paymentId) {
        return paymentRepository.findById(paymentId)
                .orElseThrow(() -> new RuntimeException("Payment not found"));
    }
    
    public Payment getPaymentByTransactionId(String transactionId) {
        return paymentRepository.findByTransactionId(transactionId)
                .orElseThrow(() -> new RuntimeException("Payment not found"));
    }
    
    public List<Payment> getPaymentsByOrder(String orderId) {
        return paymentRepository.findByOrderId(orderId)
                .map(List::of)
                .orElse(List.of());
    }
    
    public List<Payment> getPaymentsByCustomer(String customerId) {
        return paymentRepository.findByCustomerId(customerId);
    }
    
    public List<Payment> getPaymentsByStatus(Payment.PaymentStatus status) {
        return paymentRepository.findByStatus(status);
    }
    
    public List<Payment> getPaymentsByMethod(Payment.PaymentMethod method) {
        return paymentRepository.findByPaymentMethod(method);
    }
    
    public List<Payment> getAllPayments() {
        return paymentRepository.findAll();
    }
    
    public List<Payment> getPaymentsByDateRange(LocalDateTime startDate, LocalDateTime endDate) {
        return paymentRepository.findByPaymentDateBetween(startDate, endDate);
    }
    
    public Payment updatePaymentStatus(String paymentId, Payment.PaymentStatus status) {
        Payment payment = getPaymentById(paymentId);
        payment.setStatus(status);
        payment.setUpdatedAt(LocalDateTime.now());
        
        return paymentRepository.save(payment);
    }
    
    public Payment refundPayment(String paymentId, BigDecimal refundAmount, String refundReason) {
        Payment originalPayment = getPaymentById(paymentId);
        
        if (originalPayment.getStatus() != Payment.PaymentStatus.COMPLETED) {
            throw new RuntimeException("Cannot refund payment that is not completed");
        }
        
        if (refundAmount.compareTo(originalPayment.getAmount()) > 0) {
            throw new RuntimeException("Refund amount cannot exceed original payment amount");
        }
        
        // Create refund payment record
        Payment refundPayment = new Payment();
        refundPayment.setTransactionId(generateTransactionId());
        refundPayment.setOrderId(originalPayment.getOrderId());
        refundPayment.setOrderNumber(originalPayment.getOrderNumber());
        refundPayment.setCustomerId(originalPayment.getCustomerId());
        refundPayment.setCustomerName(originalPayment.getCustomerName());
        refundPayment.setCustomerEmail(originalPayment.getCustomerEmail());
        refundPayment.setAmount(refundAmount.negate()); // Negative amount for refund
        refundPayment.setPaymentMethod(originalPayment.getPaymentMethod());
        refundPayment.setGatewayResponse("Refund processed");
        refundPayment.setGatewayTransactionId(generateGatewayTransactionId());
        refundPayment.setStatus(Payment.PaymentStatus.COMPLETED);
        refundPayment.setPaymentDate(LocalDateTime.now());
        refundPayment.setRefundReason(refundReason);
        refundPayment.setRefundDate(LocalDateTime.now());
        refundPayment.setCreatedAt(LocalDateTime.now());
        refundPayment.setUpdatedAt(LocalDateTime.now());
        
        // Update original payment status
        originalPayment.setStatus(Payment.PaymentStatus.REFUNDED);
        originalPayment.setRefundReason(refundReason);
        originalPayment.setRefundDate(LocalDateTime.now());
        originalPayment.setUpdatedAt(LocalDateTime.now());
        paymentRepository.save(originalPayment);
        
        return paymentRepository.save(refundPayment);
    }
    
    private boolean processPaymentWithGateway(Payment payment) {
        // Simulate payment gateway processing
        // In a real application, this would integrate with actual payment gateways
        // like Stripe, PayPal, etc.
        
        try {
            // Simulate processing delay
            Thread.sleep(1000);
            
            // Simulate success/failure (90% success rate)
            return Math.random() > 0.1;
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return false;
        }
    }
    
    private String generateTransactionId() {
        return UUID.randomUUID().toString().replace("-", "").substring(0, 16).toUpperCase();
    }
    
    private String generateGatewayTransactionId() {
        return "GTW-" + UUID.randomUUID().toString().replace("-", "").substring(0, 12).toUpperCase();
    }
    
    public BigDecimal getTotalPayments() {
        return paymentRepository.findCompletedPayments().stream()
                .map(Payment::getAmount)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }
    
    public BigDecimal getTotalRefunds() {
        return paymentRepository.findRefundedPayments().stream()
                .map(Payment::getAmount)
                .reduce(BigDecimal.ZERO, BigDecimal::add)
                .abs();
    }
    
    public BigDecimal getTotalPaymentsByDateRange(LocalDateTime startDate, LocalDateTime endDate) {
        return paymentRepository.findByPaymentDateBetween(startDate, endDate).stream()
                .filter(payment -> payment.getStatus() == Payment.PaymentStatus.COMPLETED)
                .map(Payment::getAmount)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }
} 